File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        // 1. Fast I/O optimization to unsync C++ and C streams
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);

        maxDiameter = 0; 
        getHeight(root);
        return maxDiameter;
    }

private:
    int maxDiameter = 0;

    // 2. Pass by pointer is fine, but we can keep the memory footprint low
    int getHeight(TreeNode* root) {
        if (!root) return 0;

        int leftHeight = getHeight(root->left);
        int rightHeight = getHeight(root->right);

        maxDiameter = max(maxDiameter, leftHeight + rightHeight);

        return 1 + max(leftHeight, rightHeight);
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        
        while left <= right:
            mid = (left + right) // 2

            if nums[mid] == target:
                return mid
                
            if nums[left] <= nums[mid]:

                if nums[left] <= target < nums[mid]:
                    right = mid - 1
                else:
                    left = mid + 1
            else:
                if nums[mid] < target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
        return -1


        



"""

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low, s = 0, len(nums)
        high = s - 1

        while low <= high:
            mid = (low + high) // 2
            if (nums[mid] == target):
                return mid
            elif nums[mid] < target:
                low = mid + 1
            else:
                high = mid - 1
        return -1
        


"""